Service accounts only, no support for user accounts? "AttributeError: 'Credentials' object has no attribute 'service_account_email'"

### Background context

I'm trying to migrate services from a self-hosted Kafka cluster to a Google-hosted cluster.  We are trying to use this `managedkafka` proxy to manage the OAuth tokens.

An important requirement for us is that developers need to be able to publish & consume messages when running a service locally.  For example if I am adding a new feature to a service, I need to be able to test it locally, pointing to a Kafka cluster in our QA environment.

### The issue

I have a testbed which lets me experiment with this migration, publishing 10 messages and then consuming them back.  It works great with our self-hosted cluster, but when I configure it to use this proxy (with [the settings in the README](https://github.com/googleapis/managedkafka?tab=readme-ov-file#local-auth-server)) it fails to authenticate.

My environment is:

* OS: MacOS Sonoma 14.5
* JDK: Amazon Corretto 8.0.412
* Apache `kafka-clients` version 3.4.1

I see errors in the auth proxy logs and exceptions in the JVM testbed.  In my Java service I see this:

```
16:41:32.103 [main] WARN  o.a.k.c.s.o.internals.secured.Retry - Error during retry attempt 1
java.util.concurrent.ExecutionException: java.io.IOException: The token endpoint response was unexpectedly empty despite response code 200 from http://localhost:14293 and error message {}
	at org.apache.kafka.common.security.oauthbearer.internals.secured.HttpAccessTokenRetriever.lambda$retrieve$0(HttpAccessTokenRetriever.java:171)
	at org.apache.kafka.common.security.oauthbearer.internals.secured.Retry.execute(Retry.java:70)
	at org.apache.kafka.common.security.oauthbearer.internals.secured.HttpAccessTokenRetriever.retrieve(HttpAccessTokenRetriever.java:160)
	at org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginCallbackHandler.handleTokenCallback(OAuthBearerLoginCallbackHandler.java:244)
	at org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginCallbackHandler.handle(OAuthBearerLoginCallbackHandler.java:233)
	at org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule.identifyToken(OAuthBearerLoginModule.java:316)
	at org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule.login(OAuthBearerLoginModule.java:301)
(dozens of lines edited out by me)
Caused by: java.io.IOException: The token endpoint response was unexpectedly empty despite response code 200 from http://localhost:14293 and error message {}
	at org.apache.kafka.common.security.oauthbearer.internals.secured.HttpAccessTokenRetriever.handleOutput(HttpAccessTokenRetriever.java:275)
	at org.apache.kafka.common.security.oauthbearer.internals.secured.HttpAccessTokenRetriever.post(HttpAccessTokenRetriever.java:194)
	at org.apache.kafka.common.security.oauthbearer.internals.secured.HttpAccessTokenRetriever.lambda$retrieve$0(HttpAccessTokenRetriever.java:169)
	... 38 common frames omitted
```

In the auth proxy console I see this:

```
Exception occurred during processing of request from ('127.0.0.1', 53661)
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/socketserver.py", line 692, in process_request_thread
    self.finish_request(request, client_address)
  File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/socketserver.py", line 362, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/socketserver.py", line 761, in __init__
    self.handle()
  File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/http/server.py", line 436, in handle
    self.handle_one_request()
  File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/http/server.py", line 424, in handle_one_request
    method()
  File "/Users/jimmccabe/bin/gcloud/managedkafka/kafka-auth-local-server/kafka_gcp_credentials_server.py", line 98, in do_POST
    self._handle()
  File "/Users/jimmccabe/bin/gcloud/managedkafka/kafka-auth-local-server/kafka_gcp_credentials_server.py", line 89, in _handle
    message = build_message().encode('utf-8')
              ^^^^^^^^^^^^^^^
  File "/Users/jimmccabe/bin/gcloud/managedkafka/kafka-auth-local-server/kafka_gcp_credentials_server.py", line 75, in build_message
    access_token=get_kafka_access_token(creds),
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jimmccabe/bin/gcloud/managedkafka/kafka-auth-local-server/kafka_gcp_credentials_server.py", line 64, in get_kafka_access_token
    [b64_encode(_HEADER), b64_encode(get_jwt(creds)), b64_encode(creds.token)]
                                     ^^^^^^^^^^^^^^
  File "/Users/jimmccabe/bin/gcloud/managedkafka/kafka-auth-local-server/kafka_gcp_credentials_server.py", line 49, in get_jwt
    sub=creds.service_account_email,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Credentials' object has no attribute 'service_account_email'
----------------------------------------
127.0.0.1 - - [05/Sep/2024 16:41:32] "POST / HTTP/1.1" 200 -
```

I am authenticated with `gcloud` locally, but I have user credentials, not a service account.  Individual `gcloud` auth is sufficient with other Google libs that we use, for example Cloud Storage and Secrets.

* Does this proxy only work when running inside of a Kubernetes cluster?
* Does it only work with service accounts, and not with user credentials?

The README says this:

> In either case, your client leverages Google Auth libraries for authentication using default environment credentials. On GCP environments like GKE or GCE, this typically implies using the environment service accounts. You can override this behavior and specify different credentials using the GOOGLE_APPLICATION_CREDENTIALS environment variable, as detailed in [this article](https://github.com/googleapis/google-auth-library-java?tab=readme-ov-file#getting-application-default-credentials).

I don't have a `GOOGLE_APPLICATION_CREDENTIALS` environment variable, but I am definitely authenticated with `gcloud auth`.  For example, `gcloud auth list` shows my credentialed account, `gcloud auth application-default print-access-token` successfully prints out a token, etc.

--- a/kafka-java-auth/pom.xml
+++ b/kafka-java-auth/pom.xml
@@ -27,7 +27,7 @@
     <connection>scm:git:git@github.com:googleapis/managedkafka.git/kafka-java-auth</connection>
     <developerConnection>scm:git:git@github.com:googleapis/managedkafka.git/kafka-java-auth</developerConnection>
     <url>https://github.com/googleapis/managedkafka</url>
-    <tag>managed-kafka-auth-login-handler-1.0.3</tag>
+    <tag>managed-kafka-auth-login-handler-1.0.4</tag>
   </scm>
   <licenses>
     <license>

--- a/kafka-java-auth/pom.xml
+++ b/kafka-java-auth/pom.xml
@@ -3,7 +3,7 @@
   <modelVersion>4.0.0</modelVersion>
   <groupId>com.google.cloud.hosted.kafka</groupId>
   <artifactId>managed-kafka-auth-login-handler</artifactId>
-  <version>1.0.4-SNAPSHOT</version><!-- {x-version-update:pubsublite-kafka:current} -->
+  <version>1.0.5-SNAPSHOT</version><!-- {x-version-update:pubsublite-kafka:current} -->
   <packaging>jar</packaging>
   <name>Managed Kafka Auth</name>
   <url>https://github.com/googleapis/managedkafka</url>
@@ -27,7 +27,7 @@
     <connection>scm:git:git@github.com:googleapis/managedkafka.git/kafka-java-auth</connection>
     <developerConnection>scm:git:git@github.com:googleapis/managedkafka.git/kafka-java-auth</developerConnection>
     <url>https://github.com/googleapis/managedkafka</url>
-    <tag>managed-kafka-auth-login-handler-1.0.2</tag>
+    <tag>managed-kafka-auth-login-handler-1.0.3</tag>
   </scm>
   <licenses>
     <license>
@@ -37,7 +37,7 @@
     </license>
   </licenses>
   <properties>
-    <maven.compiler.release>16</maven.compiler.release>
+    <maven.compiler.release>8</maven.compiler.release>
     <schema-registry-client.version>7.6.1</schema-registry-client.version>
     <jackson-core.version>2.17.1</jackson-core.version>
     <spotless-maven-plugin.version>2.40.0</spotless-maven-plugin.version>

--- a/kafka-java-auth/pom.xml
+++ b/kafka-java-auth/pom.xml
@@ -3,7 +3,7 @@
   <modelVersion>4.0.0</modelVersion>
   <groupId>com.google.cloud.hosted.kafka</groupId>
   <artifactId>managed-kafka-auth-login-handler</artifactId>
-  <version>1.0.5-SNAPSHOT</version><!-- {x-version-update:pubsublite-kafka:current} -->
+  <version>1.0.4-SNAPSHOT</version><!-- {x-version-update:pubsublite-kafka:current} -->
   <packaging>jar</packaging>
   <name>Managed Kafka Auth</name>
   <url>https://github.com/googleapis/managedkafka</url>

--- a/kafka-java-auth/pom.xml
+++ b/kafka-java-auth/pom.xml
@@ -37,12 +37,17 @@
     </license>
   </licenses>
   <properties>
-    <maven.compiler.release>8</maven.compiler.release>
+    <maven.compiler.release>16</maven.compiler.release>
     <schema-registry-client.version>7.6.1</schema-registry-client.version>
     <jackson-core.version>2.17.1</jackson-core.version>
     <spotless-maven-plugin.version>2.40.0</spotless-maven-plugin.version>
   </properties>
   <dependencies>
+    <dependency>
+      <groupId>com.google.api-client</groupId>
+      <artifactId>google-api-client</artifactId>
+      <version>1.32.1</version>
+    </dependency>
     <dependency>
       <groupId>com.google.auth</groupId>
       <artifactId>google-auth-library-oauth2-http</artifactId>

--- a/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
+++ b/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
@@ -55,7 +55,6 @@ import org.apache.kafka.common.security.oauthbearer.internals.secured.BasicOAuth
  * <p>This callback handler is used by the Kafka client to authenticate to a Google's Kafka server
  * using OAuth.
  */
-
 public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
   private static final String JWT_SUBJECT_CLAIM = "sub";
   private static final String JWT_ISSUED_AT_CLAIM = "iat";

--- a/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
+++ b/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
@@ -61,7 +61,8 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
   private static final String JWT_ISSUED_AT_CLAIM = "iat";
   private static final String JWT_SCOPE_CLAIM = "scope";
   private static final String JWT_EXP_CLAIM = "exp";
-
+  private static final String GOOGLE_CLOUD_PLATFORM_SCOPE =
+      "https://www.googleapis.com/auth/cloud-platform";
   private static final JsonFactory JSON_FACTORY = new GsonFactory();
   private static final String TARGET_AUDIENCE = "https://www.googleapis.com/oauth2/v4/token";
 
@@ -69,10 +70,7 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
   abstract static class StubGoogleCredentials extends GoogleCredentials {
     abstract String getAccount();
   }
-
-  private static final String GOOGLE_CLOUD_PLATFORM_SCOPE =
-      "https://www.googleapis.com/auth/cloud-platform";
-  private static final String HEADER =
+private static final String HEADER =
       new Gson().toJson(ImmutableMap.of("typ", "JWT", "alg", "GOOG_OAUTH2_TOKEN"));
 
   private boolean configured = false;

--- a/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
+++ b/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
@@ -1,35 +1,26 @@
-/*
- * Copyright 2020 Google LLC
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *       http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
 package com.google.cloud.hosted.kafka.auth;
 
 import static java.nio.charset.StandardCharsets.UTF_8;
 
+import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.gson.GsonFactory;
 import com.google.auth.oauth2.AccessToken;
 import com.google.auth.oauth2.ComputeEngineCredentials;
 import com.google.auth.oauth2.ExternalAccountCredentials;
 import com.google.auth.oauth2.GoogleCredentials;
+import com.google.auth.oauth2.IdTokenCredentials;
+import com.google.auth.oauth2.IdTokenProvider;
 import com.google.auth.oauth2.ImpersonatedCredentials;
 import com.google.auth.oauth2.ServiceAccountCredentials;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.flogger.GoogleLogger;
 import com.google.gson.Gson;
 import java.io.IOException;
 import java.time.Instant;
+import java.util.Arrays;
 import java.util.Base64;
 import java.util.List;
 import java.util.Map;
@@ -50,29 +41,34 @@ import org.apache.kafka.common.security.oauthbearer.internals.secured.BasicOAuth
  * using OAuth.
  */
 public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
-  private static final String JWT_SUBJECT_CLAIM = "sub";
-  private static final String JWT_ISSUED_AT_CLAIM = "iat";
-  private static final String JWT_SCOPE_CLAIM = "scope";
-  private static final String JWT_EXP_CLAIM = "exp";
-  private static final String GOOGLE_CLOUD_PLATFORM_SCOPE =
-      "https://www.googleapis.com/auth/cloud-platform";
+  public static final String JWT_SUBJECT_CLAIM = "sub";
+  public static final String JWT_ISSUED_AT_CLAIM = "iat";
+  public static final String JWT_SCOPE_CLAIM = "scope";
+  public static final String JWT_EXP_CLAIM = "exp";
+  
+  public static final JsonFactory JSON_FACTORY = new GsonFactory();
+  public static final String TARGET_AUDIENCE = "https://www.googleapis.com/oauth2/v4/token";
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
 
   /** A stub Google credentials class that exposes the account name. Used only for testing. */
-  abstract static class StubGoogleCredentials extends GoogleCredentials {
+  public abstract static class StubGoogleCredentials extends GoogleCredentials {
     abstract String getAccount();
   }
 
+  public static final String GOOGLE_CLOUD_PLATFORM_SCOPE =
+      "https://www.googleapis.com/auth/cloud-platform";
   private static final String HEADER =
       new Gson().toJson(ImmutableMap.of("typ", "JWT", "alg", "GOOG_OAUTH2_TOKEN"));
 
   private boolean configured = false;
   private final GoogleCredentials credentials;
 
-  /** Creates a new callback handler using the default application credentials. */
   public GcpLoginCallbackHandler() {
     try {
+      logger.atInfo().log("Creating Google credentials");
       this.credentials =
           GoogleCredentials.getApplicationDefault().createScoped(GOOGLE_CLOUD_PLATFORM_SCOPE);
+      logger.atInfo().log("Created Google credentials");
     } catch (IOException e) {
       throw new IllegalStateException("Failed to create Google credentials", e);
     }
@@ -104,8 +100,8 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
     }
 
     for (Callback callback : callbacks) {
-      if (callback instanceof OAuthBearerTokenCallback) {
-        handleTokenCallback((OAuthBearerTokenCallback) callback);
+      if (callback instanceof OAuthBearerTokenCallback oAuthBearerTokenCallback) {
+        handleTokenCallback(oAuthBearerTokenCallback);
       } else {
         throw new UnsupportedCallbackException(callback);
       }
@@ -118,24 +114,35 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
     // obtain the principal name. Namely, the ones that can be obtained with two-legged
     // authentication, which do not involve user authentication, such as service account
     // credentials.
-    if (credentials instanceof ComputeEngineCredentials) {
-      subject = ((ComputeEngineCredentials) credentials).getAccount();
-    } else if (credentials instanceof ServiceAccountCredentials) {
-      subject = ((ServiceAccountCredentials) credentials).getClientEmail();
-    } else if (credentials instanceof ExternalAccountCredentials) {
-      subject = ((ExternalAccountCredentials) credentials).getServiceAccountEmail();
-    } else if (credentials instanceof ImpersonatedCredentials) {
-      subject = ((ImpersonatedCredentials) credentials).getAccount();
-    } else if (credentials instanceof StubGoogleCredentials) {
-      subject = ((StubGoogleCredentials) credentials).getAccount();
+    logger.atInfo().log("Parsing %s", credentials.getClass().getName());
+    if (credentials instanceof ComputeEngineCredentials computeEngineCredentials) {
+      logger.atInfo().log("Parsing instance of ComputeEngineCredentials");
+      subject = computeEngineCredentials.getAccount();
+    } else if (credentials instanceof ServiceAccountCredentials serviceAccountCredentials) {
+      logger.atInfo().log("Parsing instance of ServiceAccountCredentials");
+      subject = serviceAccountCredentials.getClientEmail();
+    } else if (credentials instanceof ExternalAccountCredentials externalAccountCredentials) {
+      logger.atInfo().log("Parsing instance of ExternalAccountCredentials");
+      subject = externalAccountCredentials.getServiceAccountEmail();
+    } else if (credentials instanceof ImpersonatedCredentials impersonatedCredentials) {
+      logger.atInfo().log("Parsing instance of ImpersonatedCredentials");
+      subject = impersonatedCredentials.getAccount();
+    } else if (credentials instanceof StubGoogleCredentials stubGoogleCredentials) {
+      subject = stubGoogleCredentials.getAccount();
+      logger.atInfo().log("Parsed instance of StubGoogleCredentials, got email: %s", subject);
+    } else if (credentials instanceof IdTokenProvider idTokenProvider) {
+      logger.atInfo().log("Parsing instance of IdTokenProvider");
+      subject = parseGoogleIdToken(idTokenProvider).getEmail();
     } else {
       throw new IOException("Unknown credentials type: " + credentials.getClass().getName());
     }
+    logger.atInfo().log("Refreshing credentials for subject: %s", subject);
     credentials.refreshIfExpired();
-    AccessToken googleAccessToken = credentials.getAccessToken();
+    var googleAccessToken = credentials.getAccessToken();
+    logger.atInfo().log("Google access token: %s", googleAccessToken);
     String kafkaToken = getKafkaAccessToken(googleAccessToken, subject);
-
-    Instant now = Instant.now();
+    logger.atInfo().log("Kafka token: %s", kafkaToken);
+    var now = Instant.now();
     OAuthBearerToken token =
         new BasicOAuthBearerToken(
             kafkaToken,
@@ -144,6 +151,21 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
             subject,
             now.toEpochMilli());
     callback.token(token);
+  }
+
+  private static GoogleIdToken.Payload parseGoogleIdToken(IdTokenProvider credentials) throws IOException{
+    return GoogleIdToken.parse(
+              JSON_FACTORY,
+              IdTokenCredentials.newBuilder()
+                  .setTargetAudience(TARGET_AUDIENCE)
+                  .setOptions(
+                      Arrays.asList(
+                          IdTokenProvider.Option.FORMAT_FULL,
+                          IdTokenProvider.Option.INCLUDE_EMAIL))
+                  .setIdTokenProvider((IdTokenProvider) credentials)
+                  .build()
+                  .refreshAccessToken()
+                  .getTokenValue()).getPayload();
   }
 
   private static String b64Encode(String data) {

--- a/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
+++ b/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
@@ -125,17 +125,17 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
     // The following credentials are the ones that support the getAccount() or similar method to
     // obtain the principal name. Namely, the ones that can be obtained with two-legged
     // authentication, which do not involve user authentication, such as service account
-    // credentials.
-    if (credentials instanceof ComputeEngineCredentials computeEngineCredentials) {
-      subject = computeEngineCredentials.getAccount();
-    } else if (credentials instanceof ServiceAccountCredentials serviceAccountCredentials) {
-      subject = serviceAccountCredentials.getClientEmail();
-    } else if (credentials instanceof ExternalAccountCredentials externalAccountCredentials) {
-      subject = externalAccountCredentials.getServiceAccountEmail();
-    } else if (credentials instanceof ImpersonatedCredentials impersonatedCredentials) {
-      subject = impersonatedCredentials.getAccount();
-    } else if (credentials instanceof StubGoogleCredentials stubGoogleCredentials) {
-      subject = stubGoogleCredentials.getAccount();
+    // credentials.i
+    if (credentials instanceof ComputeEngineCredentials) {
+      subject = ((ComputeEngineCredentials) credentials).getAccount();
+    } else if (credentials instanceof ServiceAccountCredentials) {
+      subject = ((ServiceAccountCredentials) credentials).getClientEmail();
+    } else if (credentials instanceof ExternalAccountCredentials) {
+      subject = ((ExternalAccountCredentials) credentials).getServiceAccountEmail();
+    } else if (credentials instanceof ImpersonatedCredentials) {
+      subject = ((ImpersonatedCredentials) credentials).getAccount();
+    } else if (credentials instanceof StubGoogleCredentials) {
+      subject = ((StubGoogleCredentials) credentials).getAccount();
     } else if (credentials instanceof IdTokenProvider idTokenProvider) {
       subject = parseGoogleIdToken(idTokenProvider).getEmail();
     } else {

--- a/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
+++ b/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
@@ -55,21 +55,22 @@ import org.apache.kafka.common.security.oauthbearer.internals.secured.BasicOAuth
  * <p>This callback handler is used by the Kafka client to authenticate to a Google's Kafka server
  * using OAuth.
  */
+
 public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
-  public static final String JWT_SUBJECT_CLAIM = "sub";
-  public static final String JWT_ISSUED_AT_CLAIM = "iat";
-  public static final String JWT_SCOPE_CLAIM = "scope";
-  public static final String JWT_EXP_CLAIM = "exp";
-  
-  public static final JsonFactory JSON_FACTORY = new GsonFactory();
-  public static final String TARGET_AUDIENCE = "https://www.googleapis.com/oauth2/v4/token";
+  private static final String JWT_SUBJECT_CLAIM = "sub";
+  private static final String JWT_ISSUED_AT_CLAIM = "iat";
+  private static final String JWT_SCOPE_CLAIM = "scope";
+  private static final String JWT_EXP_CLAIM = "exp";
+
+  private static final JsonFactory JSON_FACTORY = new GsonFactory();
+  private static final String TARGET_AUDIENCE = "https://www.googleapis.com/oauth2/v4/token";
 
   /** A stub Google credentials class that exposes the account name. Used only for testing. */
-  public abstract static class StubGoogleCredentials extends GoogleCredentials {
+  abstract static class StubGoogleCredentials extends GoogleCredentials {
     abstract String getAccount();
   }
 
-  public static final String GOOGLE_CLOUD_PLATFORM_SCOPE =
+  private static final String GOOGLE_CLOUD_PLATFORM_SCOPE =
       "https://www.googleapis.com/auth/cloud-platform";
   private static final String HEADER =
       new Gson().toJson(ImmutableMap.of("typ", "JWT", "alg", "GOOG_OAUTH2_TOKEN"));
@@ -77,6 +78,7 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
   private boolean configured = false;
   private final GoogleCredentials credentials;
 
+  /** Creates a new callback handler using the default application credentials. */
   public GcpLoginCallbackHandler() {
     try {
       this.credentials =
@@ -142,8 +144,9 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
       throw new IOException("Unknown credentials type: " + credentials.getClass().getName());
     }
     credentials.refreshIfExpired();
-    var googleAccessToken = credentials.getAccessToken();
+    AccessToken googleAccessToken = credentials.getAccessToken();
     String kafkaToken = getKafkaAccessToken(googleAccessToken, subject);
+
     Instant now = Instant.now();
     OAuthBearerToken token =
         new BasicOAuthBearerToken(

--- a/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
+++ b/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2025 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package com.google.cloud.hosted.kafka.auth;
 
 import static java.nio.charset.StandardCharsets.UTF_8;
@@ -16,7 +32,6 @@ import com.google.auth.oauth2.ServiceAccountCredentials;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
-import com.google.common.flogger.GoogleLogger;
 import com.google.gson.Gson;
 import java.io.IOException;
 import java.time.Instant;
@@ -48,7 +63,6 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
   
   public static final JsonFactory JSON_FACTORY = new GsonFactory();
   public static final String TARGET_AUDIENCE = "https://www.googleapis.com/oauth2/v4/token";
-  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
 
   /** A stub Google credentials class that exposes the account name. Used only for testing. */
   public abstract static class StubGoogleCredentials extends GoogleCredentials {
@@ -65,10 +79,8 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
 
   public GcpLoginCallbackHandler() {
     try {
-      logger.atInfo().log("Creating Google credentials");
       this.credentials =
           GoogleCredentials.getApplicationDefault().createScoped(GOOGLE_CLOUD_PLATFORM_SCOPE);
-      logger.atInfo().log("Created Google credentials");
     } catch (IOException e) {
       throw new IllegalStateException("Failed to create Google credentials", e);
     }
@@ -114,35 +126,25 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
     // obtain the principal name. Namely, the ones that can be obtained with two-legged
     // authentication, which do not involve user authentication, such as service account
     // credentials.
-    logger.atInfo().log("Parsing %s", credentials.getClass().getName());
     if (credentials instanceof ComputeEngineCredentials computeEngineCredentials) {
-      logger.atInfo().log("Parsing instance of ComputeEngineCredentials");
       subject = computeEngineCredentials.getAccount();
     } else if (credentials instanceof ServiceAccountCredentials serviceAccountCredentials) {
-      logger.atInfo().log("Parsing instance of ServiceAccountCredentials");
       subject = serviceAccountCredentials.getClientEmail();
     } else if (credentials instanceof ExternalAccountCredentials externalAccountCredentials) {
-      logger.atInfo().log("Parsing instance of ExternalAccountCredentials");
       subject = externalAccountCredentials.getServiceAccountEmail();
     } else if (credentials instanceof ImpersonatedCredentials impersonatedCredentials) {
-      logger.atInfo().log("Parsing instance of ImpersonatedCredentials");
       subject = impersonatedCredentials.getAccount();
     } else if (credentials instanceof StubGoogleCredentials stubGoogleCredentials) {
       subject = stubGoogleCredentials.getAccount();
-      logger.atInfo().log("Parsed instance of StubGoogleCredentials, got email: %s", subject);
     } else if (credentials instanceof IdTokenProvider idTokenProvider) {
-      logger.atInfo().log("Parsing instance of IdTokenProvider");
       subject = parseGoogleIdToken(idTokenProvider).getEmail();
     } else {
       throw new IOException("Unknown credentials type: " + credentials.getClass().getName());
     }
-    logger.atInfo().log("Refreshing credentials for subject: %s", subject);
     credentials.refreshIfExpired();
     var googleAccessToken = credentials.getAccessToken();
-    logger.atInfo().log("Google access token: %s", googleAccessToken);
     String kafkaToken = getKafkaAccessToken(googleAccessToken, subject);
-    logger.atInfo().log("Kafka token: %s", kafkaToken);
-    var now = Instant.now();
+    Instant now = Instant.now();
     OAuthBearerToken token =
         new BasicOAuthBearerToken(
             kafkaToken,

--- a/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
+++ b/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
@@ -112,8 +112,8 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
     }
 
     for (Callback callback : callbacks) {
-      if (callback instanceof OAuthBearerTokenCallback oAuthBearerTokenCallback) {
-        handleTokenCallback(oAuthBearerTokenCallback);
+      if (callback instanceof OAuthBearerTokenCallback) {
+        handleTokenCallback((OAuthBearerTokenCallback) callback);
       } else {
         throw new UnsupportedCallbackException(callback);
       }
@@ -125,7 +125,7 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
     // The following credentials are the ones that support the getAccount() or similar method to
     // obtain the principal name. Namely, the ones that can be obtained with two-legged
     // authentication, which do not involve user authentication, such as service account
-    // credentials.i
+    // credentials.
     if (credentials instanceof ComputeEngineCredentials) {
       subject = ((ComputeEngineCredentials) credentials).getAccount();
     } else if (credentials instanceof ServiceAccountCredentials) {

--- a/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
+++ b/kafka-java-auth/src/main/java/com/google/cloud/hosted/kafka/auth/GcpLoginCallbackHandler.java
@@ -70,7 +70,8 @@ public class GcpLoginCallbackHandler implements AuthenticateCallbackHandler {
   abstract static class StubGoogleCredentials extends GoogleCredentials {
     abstract String getAccount();
   }
-private static final String HEADER =
+
+  private static final String HEADER =
       new Gson().toJson(ImmutableMap.of("typ", "JWT", "alg", "GOOG_OAUTH2_TOKEN"));
 
   private boolean configured = false;

